aws/data_source_aws_iam_role.go

```diff
@@ -27,41 +28,55 @@ func dataSourceAwsIAMRole() *schema.Resource {
 				Computed: true,
 			},
 			"role_id": {
+				Type:       schema.TypeString,
+				Computed:   true,
+				Deprecated: "Use `unique_id` instead",
+			},
+			"unique_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"description": {
 				Type:     schema.TypeString,
 				Computed: true,
 			},
 			"role_name": {
+				Type:       schema.TypeString,
+				Optional:   true,
+				Deprecated: "Use `name` instead",
+			},
+			"name": {
 				Type:     schema.TypeString,
 				Required: true,
 			},
+			"create_date": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
 		},
 	}
 }
 
 func dataSourceAwsIAMRoleRead(d *schema.ResourceData, meta interface{}) error {
-	iamconn := meta.(*AWSClient).iamconn
+	name, hasName := d.GetOk("name")
+	roleName, hasRoleName := d.GetOk("role_name")
 
-	roleName := d.Get("role_name").(string)
-
-	req := &iam.GetRoleInput{
-		RoleName: aws.String(roleName),
+	if !hasName && !hasRoleName {
+		return errwrap.Wrapf("`name` must be set", nil)
```

Comment:
I don't think this is about one or the other - both have valid use cases.

As you rightly quote from errwrap "... _to wrap **errors**_ ...". 😉  We don't have a real error (as in nothing matching [`error` interface](https://blog.golang.org/error-handling-and-go#TOC_2.)) as we're literally constructing new string-based one here. errwrap is meant to be used for cases where you get a typed error (e.g. from AWS SDK) and you want to add some extra context without hiding the original error behind.

`fmt.Errorf("Extra context: %s", origErr)` comes back as `&errors.errorString{s:"Extra context: stringified original error"}`. You have no way to reliably figure out what the original error was, except by doing string match of the message. Errwrap provides way to do this: https://github.com/hashicorp/errwrap#custom-types

We'd rarely use `errwrap` when returning error from CRUD, because there's unlikely going to be anything relying on the error type in the rest of the stack. We only present it to the user as text.

Let me know if that explanation makes sense.